### DISCUSSION TITLE
Quarkus now supports test profiles

### DISCUSSION THREAD

**Discussion by giscus** [quarkusio/quarkus#27696] [Quarkus Blog/Website]
# Quarkus now supports test profiles

Quarkus: Supersonic Subatomic Java

https://quarkus.io/blog/quarkus-test-profiles/

**Comment by WesternGun** [comment]
Test profile should always have class and constructor level public visibility, as for 2.7.5, for Quarkus to find the constructor. It can be empty.

**Reply by WesternGun** [reply]
Test profile properties is not interchangeable with YAML format config. This can cause problem when I want to put YAML format config here. For example, if YAML config has not levels containing `.`, I can define `aa.bb.cc` here in test profile and will be interpreted as  3 levels as YAML; but `aa.b.b.cc` will cause parsing error. With or without quotes around `b.b`, it will fail(in different places). Maybe a solution can be defining a separate profile with same name as the return value of `getConfigProfile()`.

**Reply by WesternGun** [reply]
A correction: my mistake in last post, test profiles DOES support quotations in properties keys, i.e., you can use `""` to mark string containing `.` as a single level in `getConfigOverrides()` method, the same way YAML does. For example, `org."1.2.3".id` are recognized as a valid 3 level YAML structure. Cannot edit/delete last post so append here. My apologies.

**Reply by WesternGun** [reply]
... but you will need escaping, of course.

```
public Map<String, String> getConfigOverrides() {
    return Map.of(
        "org.\"1.2.3\".id", "foo"
    );
}
```

**Comment by Franknjava** [comment]
Hi,

I do have a question to TestProfiles and overriding config, because I do not quite get it to work, at least not in an elegant way. I have a common TestProfile which produces several beans used during the tests like:

```
public class TestConfig {

    @ApplicationScoped
    @Produces
    @Named("bean1")
    public Bean1 bean1() {
        return new Bean1();
    }

    @ApplicationScoped
    @Produces
    @Named("bean2")
    public Bean2 bean2() {
        return new Bean2();
    }

    @ApplicationScoped
    @Produces
    @Named("bean3")
    public Bean3 bean3() {
        return new Bean3();
    }
}
```

Now, for one test I want one bean to be replaced with a different implementation. Therefore, I added a profile for that special test like this:

```
public class SpecialTestProfile implements QuarkusTestProfile {
    @ApplicationScoped
    @Produces
    @Named("bean2")
    Bean2 bean2() {
        return new Bean2Alternative();
    }
}
```

If I run this, I get ambiguous bean exception. I was able to fix this by setting the profile name explicitely in SpecialTestProfile like this:

```
    @Override
    public String getConfigProfile() {
        return this.getClass().getSimpleName();
    }
```

Afterward I added @IfBuildProfile annotation to SpecialTestProfile and @UnlessBuildProfile annotation to TestConfig. Both annotations were needed to avoid ambiguous bean exception.

This solution is not very nice and the certainly is a cleaner way to do this. I also tried @Alternative annotation on SpecialTestProfile but it did not work.

Any insights on this would be welcome.
